invenio/modules/records/access.py

```diff
@@ -133,7 +163,7 @@ def check_user_can_view_record(user_info, recid):
         return (0, '')
 
     restricted_collections = get_restricted_collections_for_recid(
-        recid, recreate_cache_if_needed=False
+        recid, recreate_cache_if_needed=True
```

Comment:
I just had an extra look without running the code, but as far as I remember, it's when you upload new records. First bibupload runs, and part of this fix ensures that a record which is in the root collection can be displayed even though webcoll haven't run. Now webcoll runs and assigns the record to a multiple restricted collections (the restricted root collection is authorised to guest).
Now you try to display the record, which means that ``check_user_can_view_record()`` is called. Without recreating the cache, the root collection cache does not contain the new record (even though webcoll have just assigned it). The other collection caches however do contain the record. The result is that the root collection is not included in the list of restricted collections that the record is part of, which means that a guest user (who is only authorised to the root restricted collection) will not be authorised to see the record because he/she doesn't have access to the other restricted collections.

All in all, this is a symptom fix for a much bigger issue of the underlying authorisation system and cache and search.